onnxruntime/core/graph/graph_utils.cc

```diff
@@ -191,51 +137,221 @@ static void UpdateImplicitInputNameInSubgraph(Node& node,
-  const int input_edge_dst_arg = input_edge.GetSrcArgIndex();
-
-  const Node::EdgeEnd& output_edge = *node.OutputEdgesBegin();
-  const Node& output_edge_node = output_edge.GetNode();
-  const NodeIndex output_edge_node_index = output_edge_node.Index();
-  const int output_edge_dst_arg = output_edge.GetDstArgIndex();
-
-  // check if the output from the node to remove is implicit input to the downstream node.
-  // if so, it's used in subgraph in that node and we have to check if it's valid to update that subgraph
-  if (OutputEdgeProvidesImplicitInput(output_edge)) {
-    // the output we need to wire to the downstream node, and the output name from the node we want to remove
-    const auto& output_name = GetNodeOutputName(input_edge_node, output_edge.GetSrcArgIndex());
-    const auto& removed_output_name = GetNodeOutputName(node, 0);
-
-    Node& mutable_output_edge_node = *graph.GetNode(output_edge_node_index);
-
-    if (CanUpdateImplicitInputNameInSubgraph(mutable_output_edge_node, removed_output_name, output_name)) {
-      UpdateImplicitInputNameInSubgraph(mutable_output_edge_node, removed_output_name, output_name);
-    } else {
-      // we can't safely remove this node
-      return false;
+  return output_edges;
+}
+
+/** Removes a set of GraphEdges from the graph. */
+static void RemoveOutputEdges(Graph& graph, const std::vector<GraphEdge>& edges) {
+  for (const auto& edge_to_remove : edges) {
+    graph.RemoveEdge(edge_to_remove.src_node,
+                     edge_to_remove.dst_node,
+                     edge_to_remove.src_arg_index,
+                     edge_to_remove.dst_arg_index);
+  }
+}
+
+/** Removes a node with a single incoming node. */
+static bool RemoveNodeWithSingleNodeIn(Graph& graph, Node& node) {
+  // Store info for input and output edges.
+  std::vector<GraphEdge> output_edges = GetNodeOutputEdges(node);
+  const Node::EdgeEnd& input_edge_end = *node.InputEdgesBegin();
+  const auto& input_edge_node = input_edge_end.GetNode();
+  const GraphEdge input_edge{input_edge_node.Index(),
+                             node.Index(),
+                             input_edge_end.GetSrcArgIndex(),
+                             input_edge_end.GetDstArgIndex(),
+                             GetNodeOutputName(input_edge_node, input_edge_end.GetSrcArgIndex())};
+
+  // Check if one of the edges provides an implicit input to a subgraph, and if so, check if the subgraph allows
+  // the node to be safely removed.
+  for (const auto& output_edge : output_edges) {
+    if (OutputEdgeProvidesImplicitInput(graph, output_edge)) {
+      Node& mutable_output_edge_node = *graph.GetNode(output_edge.dst_node);
+      if (!CanUpdateImplicitInputNameInSubgraph(mutable_output_edge_node, output_edge.arg_name, input_edge.arg_name)) {
+        LOGS_DEFAULT(WARNING) << " Removal of node " << node.Name()
+                              << " failed due to problem in updating a subgraph.";
```

Comment:
>problem in [](start = 49, length = 10)

maybe 'as we are not able to update the subgraph' would be clearer? if I read this it sounds like changes were made to the subgraph, which is a bit scarier. :)

Reply:
Yep -- rephrased that after the refactoring to a separate method.